hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcConnection.java

```diff
@@ -321,31 +320,39 @@ public void run(boolean cancelled) throws IOException {
         } else {
           Channel channel = channelRef.get();
           if (channel == null) {
-            try {
-              Channel newChannel = connect();
-              if (!channelRef.compareAndSet(null, newChannel)) {
-                newChannel.close();
+            connect().addListener(new ChannelFutureListener() {
+              @Override public void operationComplete(ChannelFuture channelFuture) {
+                Channel ch = channelFuture.channel();
```

Comment:
What happens if the future is unsuccessful?

Reply:
Ok, let me rephrase this: won't `channelFuture.channel()` throw an exception? You call it, but you don't use the returned channel when the future is not successful.

Reply:
Looking at the implementation of channel() it only returns a property, so not throwing any exception. Also same pattern: https://github.com/apache/hbase/blob/73ec449dccc7519b8b27a5aae0bd222e4aeb42d1/hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcConnection.java#L282

Reply:
Great, thanks. I still don't understand why we call `channelFuture.channel()` if we don't use it's return value. We don't use it if `channelFuture.isSuccess() == false`.

Reply:
I don't see if it makes any difference, but will change it in a later commit. Let's focus on the other question, because we running pointless circles here.